### PR TITLE
Remove WhileQ

### PR DESCRIPTION
**Pull request by ddn0**
commit 4a636e2ea7a6809775e528e598183b43127db56f (HEAD -> feature/whileq, ddn0/feature/whileq)
Author: Donald Nguyen <ddn0@users.noreply.github.com>
Date:   Fri Oct 1 14:25:00 2021 -0700

    libgalois: Remove unused serial specializations of for_each

    According to benchmarking (below), serial specializations (WhileQ,
    SerFIFO, SerLIFO) of katana::for_each could be valuable, though no
    application currently uses these specializations as they are always
    instantiated with their concurrent variants.

    In any event, the interface to these serial specializations is not
    appropriate. Both katana::do_all and katana::for_each already support
    compile-time traits and configurable scheduling policies; if
    specialization is needed, it should be done through these traits rather
    than wrapping do_all and for_each themselves in another layer of
    abstraction.

    Given that WhileQ, SerFIFO and SerLIFO are not used by current code,
    simply remove them and anticipate a more idiomatic interface to serial
    specialization if it is deemed broadly useful.

    The following are benchmark results for executing a trivial function
    using std::for_each, katana::do_all (1 and 4 threads), katana::for_each
    (1 and 4 threads). The overhead of do_all is quite modest compared to
    std::for_each. The main overhead seems to be waiting for a barrier, and
    this cost diminishes for larger inputs.

    The cost of katana::for_each is high relative to std::for_each,
    presumably because the former copies (and allocates) its input while
    std::for_each is simply incrementing a loop variable.

      ddn@spork:katana-conda$ external/katana/libgalois/test/foreach-bench-test
      2021-10-01T14:28:05-07:00
      Running external/katana/libgalois/test/foreach-bench-test
      Run on (12 X 4500 MHz CPU s)
      CPU Caches:
        L1 Data 32 KiB (x6)
        L1 Instruction 32 KiB (x6)
        L2 Unified 256 KiB (x6)
        L3 Unified 12288 KiB (x1)
      Load Average: 3.87, 5.40, 3.71
      ***WARNING*** CPU scaling is enabled, the benchmark real time measurements may be noisy and will incur extra overhead.
      --------------------------------------------------------------------------------
      Benchmark                      Time             CPU   Iterations UserCounters...
      --------------------------------------------------------------------------------
      StdForEach/1024              603 ns          603 ns      1252562 items_per_second=1.69812G/s
      StdForEach/65536           43048 ns        43048 ns        15302 items_per_second=1.52241G/s
      StdForEach/1048576        769944 ns       769945 ns          918 items_per_second=1.36188G/s
      DoAll/1024                  6024 ns         6023 ns       126948 items_per_second=170.004M/s
      DoAll/65536                19906 ns        19905 ns        34735 items_per_second=3.29245G/s
      DoAll/1048576             236775 ns       236773 ns         2900 items_per_second=4.42861G/s
      SerialDoAll/1024             754 ns          754 ns       874388 items_per_second=1.35811G/s
      SerialDoAll/65536          46796 ns        46796 ns        14951 items_per_second=1.40047G/s
      SerialDoAll/1048576       761156 ns       761114 ns          894 items_per_second=1.37769G/s
      ForEach/1024               16009 ns        16009 ns        47351 items_per_second=63.9644M/s
      ForEach/65536             628142 ns       628128 ns         1113 items_per_second=104.335M/s
      ForEach/1048576          9995756 ns      9995763 ns           72 items_per_second=104.902M/s
      SerialForEach/1024          7496 ns         7496 ns        88610 items_per_second=136.609M/s
      SerialForEach/65536       448630 ns       448628 ns         1551 items_per_second=146.081M/s
      SerialForEach/1048576    7989831 ns      7989828 ns           78 items_per_second=131.239M/s

### CODE REVIEW

**Review by amberhassaan — APPROVED** [review]
why is `std::for_each` so much faster? At first I thought it was using a `parallel policy`, but seems like it's just serial.

File: libgalois/include/katana/Loops.h [hunk 1]
```diff
@@ -170,6 +169,10 @@ struct StdForEach {
   }
 };
 
+/**
+ * Helper functor to invoke katana::for_each with the same interface as
+ * katana::do_all
+ */
 struct ForEach {
```
Comment (roshandathathri) [inline libgalois/include/katana/Loops.h:176]: The comment is confusing. This provides the same interface as `DoAll`, not `do_all` right?

File: libgalois/test/foreach-bench.cpp [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+
+  for (auto _ : state) {
+    RunForEach(input, output);
+  }
+
+  VerifyOutput(output);
+  state.SetItemsProcessed(state.iterations() * size);
+}
+
+BENCHMARK(StdForEach)->Apply(MakeArguments);
```
Comment (roshandathathri) [inline libgalois/test/foreach-bench.cpp:148]: So the time includes time to make input and output?

**Review by roshandathathri — APPROVED** [review]
LGTM

File: libgalois/test/foreach-bench.cpp (hunk 2 above)
Comment (ddn0) [reply libgalois/test/foreach-bench.cpp:148]: The timed region is defined by `for (auto _ : state) { ... }` so doesn't include the time to make `input` or `output` or to validate `output`.

**Comment by ddn0** [comment]
> why is `std::for_each` so much faster? At first I thought it was using a `parallel policy`, but seems like it's just serial.

`std::for_each` is faster than `katana::do_all` (threads: 4) because it doesn't have to wait for a thread barrier.
```
StdForEach/1024              603 ns          603 ns      1252562 items_per_second=1.69812G/s
DoAll/1024                  6024 ns         6023 ns       126948 items_per_second=170.004M/s
```

For `std::for_each` versus `katana::do_all` (threads: 1), the gap is smaller and it is somewhat odd that we don't have exact parity between `std::for_each` and `katana::do_all` but the performance difference diminishes for larger datasets.
```
StdForEach/1024              603 ns          603 ns      1252562 items_per_second=1.69812G/s
StdForEach/65536           43048 ns        43048 ns        15302 items_per_second=1.52241G/s
StdForEach/1048576        769944 ns       769945 ns          918 items_per_second=1.36188G/s
SerialDoAll/1024             754 ns          754 ns       874388 items_per_second=1.35811G/s
SerialDoAll/65536          46796 ns        46796 ns        14951 items_per_second=1.40047G/s
SerialDoAll/1048576       761156 ns       761114 ns          894 items_per_second=1.37769G/s
```

File: libgalois/test/foreach-bench.cpp (hunk 2 above)
Comment (roshandathathri) [reply libgalois/test/foreach-bench.cpp:148]: Thanks. That's good to know. @danielmawhirter

**Pull request merged**